// instantsearch.js/cjs/components/SmartSort/SmartSort.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _preact = require("preact");

var _Template = _interopRequireDefault(require("../Template/Template"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** @jsx h */
var SmartSort = function SmartSort(_ref) {
  var cssClasses = _ref.cssClasses,
      templates = _ref.templates,
      isSmartSorted = _ref.isSmartSorted,
      isVirtualReplica = _ref.isVirtualReplica,
      refine = _ref.refine;
  return isVirtualReplica ? (0, _preact.h)("div", {
    className: cssClasses.root
  }, (0, _preact.h)(_Template.default, {
    templateKey: "text",
    templates: templates,
    rootProps: {
      className: cssClasses.text
    },
    data: {
      isSmartSorted: isSmartSorted
    }
  }), (0, _preact.h)("button", {
    type: "button",
    className: cssClasses.button,
    onClick: function onClick() {
      if (isSmartSorted) {
        refine(0);
      } else {
        refine(undefined);
      }
    }
  }, (0, _preact.h)(_Template.default, {
    rootTagName: "span",
    templateKey: "button",
    templates: templates,
    data: {
      isSmartSorted: isSmartSorted
    }
  }))) : null;
};

var _default = SmartSort;
exports.default = _default;
